packages/sdk/src/store/store.ts

```diff
@@ -257,32 +264,81 @@ export class Store implements IStore {
     };
   }
 
-  private async getPeerToUse(): Promise<PeerId | undefined> {
-    let peerId: PeerId | undefined;
+  private async getPeerToUse(pubsubTopic: string): Promise<PeerId | undefined> {
+    const peers = await this.filterConnectedPeers(pubsubTopic);
+
+    const peer = this.options.peers
+      ? await this.getPeerFromConfigurationOrFirst(peers, this.options.peers)
+      : peers[0]?.id;
+
+    return peer;
+  }
 
-    if (this.options?.peer) {
-      const connectedPeers = await this.connectionManager.getConnectedPeers();
+  private async getPeerFromConfigurationOrFirst(
+    peers: Peer[],
+    configPeers: string[]
+  ): Promise<PeerId | undefined> {
+    const storePeers = configPeers.map(multiaddr);
+    const missing = [];
 
-      const peer = connectedPeers.find(
-        (p) => p.id.toString() === this.options?.peer
+    for (const peer of storePeers) {
+      const matchedPeer = peers.find(
+        (p) => p.id.toString() === peer.getPeerId()?.toString()
       );
-      peerId = peer?.id;
 
-      if (!peerId) {
+      if (matchedPeer) {
+        return matchedPeer.id;
+      }
+
+      missing.push(peer);
+    }
+
+    while (missing.length) {
+      const toDial = missing.pop();
+
+      if (!toDial) {
+        return;
+      }
+
+      try {
+        const c = await this.libp2p.dial(toDial);
+
+        if (c) {
+          return peerIdFromString(toDial.getPeerId() as string);
+        }
+      } catch (e) {
         log.warn(
-          `Passed node to use for Store not found: ${this.options.peer}. Attempting to use random peers.`
+          `Failed to dial peer from options.peers list for Store protocol. Peer:${toDial.getPeerId()}, error:${e}`
         );
       }
     }
 
-    const peerIds = this.peerManager.getPeers();
+    log.warn(
+      `Passed node to use for Store not found: ${configPeers.toString()}. Attempting to use first available peers.`
+    );
+
+    return peers[0]?.id;
+  }
+
+  private async filterConnectedPeers(pubsubTopic: string): Promise<Peer[]> {
```

Comment:
I feel like this function should be generalized to check any protocol(s) and made part of connection manager. Filtering peers for specified shard/pubsub and codec sounds like out of responsibility of store class

Reply:
good point and I am trying to achieve that in https://github.com/waku-org/js-waku/pull/2425